tsdb/shard.go

```diff
@@ -252,19 +491,22 @@ func (s *Shard) ValidateAggregateFieldsInStatement(measurementName string, stmt
 
 // deleteSeries deletes the buckets and the metadata for the given series keys
 func (s *Shard) deleteSeries(keys []string) error {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
 	if err := s.db.Update(func(tx *bolt.Tx) error {
 		b := tx.Bucket([]byte("series"))
 		for _, k := range keys {
 			if err := b.Delete([]byte(k)); err != nil {
 				return err
 			}
-			if err := tx.DeleteBucket([]byte(k)); err != nil {
+			if err := tx.DeleteBucket([]byte(k)); err != nil && err != bolt.ErrBucketNotFound {
```

Comment:
It's definitely good to know the return errors from Bolt's operations. They're generally [documented in the godoc](http://godoc.org/github.com/boltdb/bolt#Bucket.DeleteBucket) but you can also find the exact list of named errors if you open up the function:

https://github.com/boltdb/bolt/blob/master/bucket.go#L208-L212
https://github.com/boltdb/bolt/blob/master/bucket.go#L218-L223

Failing from a `bolt.ErrBucketNotFound` and returning that error causes the whole `Update()` to rollback which is probably not what you want.